File: 0086-partition-list/0086-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, val=0, next=None):
#     self.val = val
#     self.next = next
class Solution:
  def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
    if head is None:
      return head
    left, right = ListNode(), ListNode()
    curr, leftTail, rightTail = head, left, right
    nextNode = None
    while curr:
      if curr.val < x:
        leftTail.next = curr
        leftTail = leftTail.next
      else:
        rightTail.next = curr
        rightTail = rightTail.next
      nextNode = curr.next
      curr.next = None
      curr = nextNode
    leftTail.next = right.next
    return left.next

File: 0086-partition-list/0086-partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *   int val;
 *   ListNode next;
 *   ListNode() {}
 *   ListNode(int val) { this.val = val; }
 *   ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode partition(ListNode head, int x) {
    if (head == null) return head;
    ListNode left = new ListNode();
    ListNode right = new ListNode();
    ListNode curr = head;
    ListNode leftTail = left;
    ListNode rightTail = right;
    ListNode nextNode;
    while (curr != null) {
      if (curr.val < x) {
        leftTail.next = curr;
        leftTail = leftTail.next;
      } else {
        rightTail.next = curr;
        rightTail = rightTail.next;
      }
      nextNode = curr.next;
      curr.next = null;
      curr = nextNode;
    }
    leftTail.next = right.next;
    return left.next;
  }
}